coriolis/providers/backup_writers.py

```diff
@@ -213,29 +220,98 @@ def _send_msg(self, data):
                 raise exception.CoriolisException(
                     "write_data exited with error code %r (%s)" % (
                         ret_val, _WRITER_ERR_MAP.get(int(ret_val))))
-        self._msg_id += 1
+
         self._stdin.write(data)
         self._stdin.flush()
         self._stdout.read(4)
 
     def _open(self):
         self._exec_helper_cmd()
+        self._sender_evt = eventlet.spawn(
+            self._sender)
+        self._encoder_evt = [
+            eventlet.spawn(self._encoder),
+            eventlet.spawn(self._encoder),
+            eventlet.spawn(self._encoder)]
 
     def seek(self, pos):
         self._offset = pos
 
     def truncate(self, size):
         pass
 
+    def _sender(self):
+        while True:
+            data = self._sender_q.get()
+            try:
+                self._send_msg(data)
+            except Exception as err:
+                self._exception = err
+                raise
+            finally:
+                self._sender_q.task_done()
+
+    def _encoder(self):
+        while True:
+            payload = self._enc_q.get()
+            try:
+                data = self._encode_data(
+                    payload["data"],
+                    payload["offset"],
+                    payload["msg_id"])
+                self._sender_q.put(data)
+            except Exception as err:
+                self._exception = err
+                raise
+            finally:
+                self._enc_q.task_done()
+
     def write(self, data):
-        self._send_msg(self._encode_data(data))
+        if self._closing:
+            raise exception.CoriolisException(
+                "Attempted to write to a closed writer.")
+
+        if self._exception:
+            raise exception.CoriolisException(
+                    "Failed to write data. See log "
+                    "for details.") from self._exception
+
+        payload = {
+            "offset": self._offset,
+            "data": data,
+            "msg_id": self._msg_id,
+        }
+        self._enc_q.put(payload)
         self._offset += len(data)
+        self._msg_id += 1
+
+    def _wait_for_queues(self):
+        LOG.info("Waiting for unfinished transfers to complete")
+        while (self._enc_q.unfinished_tasks or
```

Comment:
I'd add a reasonable timeout here in case either we or eventlet screws up and this hangs.

Reply:
Suggestions?

Reply:
Considereing we only call this "on the way out" before closing, I'd expect 120s would strike a good balance between and not too long, my only concern is that it doesn't somehow hang forever due to an edge-case

Reply:
set a 600 second timeout. Just to account for really slow networks. If need be, we can lower it later.